File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.py
'''
For node at i'th position position of node.left will be 2*i 
and  node.right will be 2*i + 1. We traverse through the tree
and assign position to nodes. Width of a level is difference of 
positions between the leftmost and rightmost nodes in that level.
'''
class Solution(object):
    def widthOfBinaryTree(self, root):
        if not root: return 0
        q = [(root, 1)]
        ans = 1  # because cases like [1], [1,3] etc. has width 1
        while q:            
		# first element in q is leftmost element in that level and
		# last element is rightmost element
		# the difference in their position is the width of that level
            if len(q) >= 2:
                ans = max(q[-1][1] - q[0][1] + 1, ans)
            new_q = []          # this stores nodes for next level  
            while q:
                node, pos = q.pop(0)                      
                if node.left: new_q.append((node.left, 2*pos))
                if node.right: new_q.append((node.right, 2*pos + 1))            
            q = new_q
        return ans

File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
int widthOfBinaryTree(TreeNode * root) {
  if (!root) return 0;
  int ans = 0;
  queue <pair<TreeNode*,long long int>> q;
  q.push({ root,0ll });
  while(!q.empty()){
    int size = q.size();
    long long int curMin = q.front().second;
    int leftMost, rightMost;
    for (int i=0; i<size; i++) {
      long long int cur_id = q.front().second - curMin; // subtracted to prevent integer overflow
      TreeNode * temp = q.front().first;
      q.pop();
      if (i == 0) leftMost = cur_id;
      if (i == size-1) rightMost = cur_id;
      if(temp->left)  q.push({ temp->left, cur_id*2 + 1 });
      if(temp->right) q.push({ temp->right, cur_id*2 + 2 });
    }
    ans = max(ans, rightMost-leftMost+1);
  }
  return ans;
}
};